### PR TITLE
Add Logstash to rpc-extras

### PR DESCRIPTION
**Pull request by andymcc**
Add the logstash role and setup to rpc-extras
* Converted from openstack-ansible-deployment juno branch
* Reworked some plays for formatting
* Moved the logstash restart to be a handler
* split into 4 tasks:
  * pre-install
  * install
  * post-install
  * logsatsh-web-disable

Note that the pip configuration does not use the locked-down
os-ansible-deployment python repository as there is currently no simple
mechanism to include the required python wheels for rpc-extras.

### CODE REVIEW

File: playbooks/roles/logstash/defaults/main.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+logstash_apt_keys:
+  - { url: "http://packages.elasticsearch.org/GPG-KEY-elasticsearch", state: "present" }
+
+logstash_apt_packages:
+  - logstash
+  - logstash-contrib
+  - openjdk-7-jre
+
+logstash_pip_packages:
+  - python-memcached
```
Comment (odyssey4me) [inline playbooks/roles/logstash/defaults/main.yml]: Are these pip packages actually required? To my knowledge neither are actually used.

**Comment by odyssey4me** [comment]
@andymcc I'd strongly suggest adding https://review.openstack.org/165546 into this patch for now. We can add multiline filtering a little later, but without removing multiline and increasing the workers logstash will be pretty useless as a rsyslog receiver and processor.

**Comment by andymcc** [comment]
@odyssey4me ahh - i tried to get clever and take the version that was the "last" in master before it got removed. I had initially taken it from juno, so moved back to do that! Should be good to go now.

File: playbooks/roles/logstash/defaults/main.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+logstash_apt_repo_url: "http://packages.elasticsearch.org/logstash/1.4/debian"
+
+logstash_apt_repos:
+  - { repo: "deb {{ logstash_apt_repo_url }} stable main", state: "present" }
+
+logstash_apt_keys:
+  - { url: "http://packages.elasticsearch.org/GPG-KEY-elasticsearch", state: "present" }
+
+logstash_port: 5544
+elasticsearch_tcp_port: 9300
```
Comment (odyssey4me) [inline playbooks/roles/logstash/defaults/main.yml]: It'd be nice to have comments about what these settings are for. Not critical, but nice. :)

**Comment by odyssey4me** [comment]
LGTM - I haven't yet tested it, but it looks like it should work.

File: playbooks/roles/logstash/defaults/main.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+elasticsearch_mode: transport
+elasticsearch_cluster: openstack
+
+logstash_apt_packages:
+  - logstash
+  - logstash-contrib
+  - openjdk-7-jre
+
+logstash_pip_packages:
```
Comment (odyssey4me) [inline playbooks/roles/logstash/defaults/main.yml]: this can be removed entirely

File: playbooks/roles/logstash/meta/main.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  min_ansible_version: 1.6.6
+  platforms:
+    - name: Ubuntu
+      versions:
+        - trusty
+  categories:
+    - rackspace
+    - logstash
+
+dependencies:
```
Comment (odyssey4me) [inline playbooks/roles/logstash/meta/main.yml]: this can be removed entirely

File: playbooks/roles/logstash/tasks/logstash_install.yml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  register: install_apt_packages
+  until: install_apt_packages|success
+  retries: 5
+  delay: 2
+  with_items: logstash_apt_packages
+  tags:
+    - logstash-apt-packages
+    - logstash-install
+
+- name: Install pip packages
```
Comment (odyssey4me) [inline playbooks/roles/logstash/tasks/logstash_install.yml]: this can be removed entirely

**Comment by odyssey4me** [comment]
I've tested with all the pip stuff removed and it works for me.

File: playbooks/roles/logstash/defaults/main.yml [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+logstash_apt_repo_url: "http://packages.elasticsearch.org/logstash/1.4/debian"
+
+logstash_apt_repos:
+  - { repo: "deb {{ logstash_apt_repo_url }} stable main", state: "present" }
+
+logstash_apt_keys:
+  - { url: "http://packages.elasticsearch.org/GPG-KEY-elasticsearch", state: "present" }
+
+logstash_port: 5544
```
Comment (odyssey4me) [inline playbooks/roles/logstash/defaults/main.yml]: logstash_port should be renamed to something like logstash_syslog_listen_port or something like that. It need to be more descriptive of its purpose.

**Comment by andymcc** [comment]
I removed the pip_dependency tasks/dependency & vars.
I'm not sure about renaming the port - its not a massive change but its inconsistent (most services aren't named that specifically), and logstash only has 1 listen port here, its not explicitly for syslog either.

File: README.md [hunk 7]
```diff
@@ -55,5 +56,7 @@ openstack-ansible setup-everything.yml
 # Ansible Roles
 
 * `horizon_extensions`
+* `logstash`
+* `module_import`
```
Comment (git-harry) [inline README.md]: What's module_import?

File: playbooks/roles/logstash/defaults/main.yml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+logstash_apt_repo_url: "http://packages.elasticsearch.org/logstash/1.4/debian"
+
+logstash_apt_repos:
+  - { repo: "deb {{ logstash_apt_repo_url }} stable main", state: "present" }
+
+logstash_apt_keys:
+  - { url: "http://packages.elasticsearch.org/GPG-KEY-elasticsearch", state: "present" }
```
Comment (git-harry) [inline playbooks/roles/logstash/defaults/main.yml]: https?

File: README.md (hunk 7 above)
Comment (odyssey4me) [inline README.md]: It's since been removed from the master branch and this line is not part of the PR.

File: playbooks/roles/logstash/defaults/main.yml (hunk 8 above)
Comment (odyssey4me) [inline playbooks/roles/logstash/defaults/main.yml]: apt is not configured to be able to use https-based repositories by default. Are you suggesting the repository be https or just the key URL?

File: playbooks/roles/logstash/defaults/main.yml [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+logstash_port: 5544
+elasticsearch_tcp_port: 9300
+elasticsearch_mode: transport
+elasticsearch_cluster: openstack
+
+logstash_apt_packages:
+  - logstash
+  - logstash-contrib
+  - openjdk-7-jre
```
Comment (odyssey4me) [inline playbooks/roles/logstash/defaults/main.yml]: `openjdk-7-jre-headless` can be used for a slightly smaller footprint.

**Comment by odyssey4me** [comment]
:+1: I'm happy for this to be merged.

**Comment by odyssey4me** [comment]
:+1: tested and confirmed that logstash deploys as expected and is able to communicate with elasticsearch
